src/EventStore.Core/Services/Transport/Grpc/Redaction.SwitchChunk.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Threading.Tasks;
+using EventStore.Client.Redaction;
+using EventStore.Core.Data.Redaction;
+using EventStore.Core.Messages;
+using EventStore.Core.Messaging;
+using EventStore.Plugins.Authorization;
+using Grpc.Core;
+
+namespace EventStore.Core.Services.Transport.Grpc {
+	internal partial class Redaction {
+		private static readonly Operation SwitchChunkOperation = new(EventStore.Plugins.Authorization.Operations.Node.Redaction.SwitchChunk);
+
+		public override async Task SwitchChunk(
+			IAsyncStreamReader<SwitchChunkReq> requestStream,
+			IServerStreamWriter<SwitchChunkResp> responseStream,
+			ServerCallContext context) {
+
+			var user = context.GetHttpContext().User;
+			if (!await _authorizationProvider.CheckAccessAsync(user, SwitchChunkOperation, context.CancellationToken).ConfigureAwait(false))
+				throw RpcExceptions.AccessDenied();
+
+			await SwitchChunksLock().ConfigureAwait(false);
+			try {
+				await SwitchChunks(context, requestStream, responseStream).ConfigureAwait(false);
+			} finally {
+				await SwitchChunksUnlock().ConfigureAwait(false);
+			}
+		}
+
+		private async Task SwitchChunksLock() {
+			var lockTcs = new TaskCompletionSource<RedactionMessage.SwitchChunkLockCompleted>();
+			_bus.Publish(new RedactionMessage.SwitchChunkLock(
+				new CallbackEnvelope(msg => lockTcs.SetResult(msg as RedactionMessage.SwitchChunkLockCompleted))
+			));
+
+			var completionMsg = await lockTcs.Task.ConfigureAwait(false);
+
+			// it's important not to throw any exceptions beyond this point in this method
+			// (except to indicate that acquiring the lock failed), otherwise we can end up
+			// in a situation where the lock remains permanently acquired.
+
+			if (completionMsg is null) // should never happen
+				return;
+
+			var result = completionMsg.Result;
+			if (result != SwitchChunkLockResult.Success)
+				throw RpcExceptions.RedactionLockFailed();
+		}
+
+		private async Task SwitchChunks(
+			ServerCallContext context,
+			IAsyncStreamReader<SwitchChunkReq> requestStream,
+			IServerStreamWriter<SwitchChunkResp> responseStream) {
+
+			// write empty headers to indicate that acquiring the lock was successful
+			await context.WriteResponseHeadersAsync(new Metadata()).ConfigureAwait(false);
```

Comment:
i dont mind particularly, but our of curiosity is this better as a header than it would have been as a message?

Reply:
actually, the comment may be a bit misleading:
to be able to throw an RPC exception before any message has been sent on the channel, we need to send the headers (that will contain the exception details in the metadata)

i'll rephrase it.

i like it mainly for its simplicity:
- the proto file is simpler
- locking is a pre-requisite to be able to switch chunks so, using the headers fits this purpose well

Reply:
the headers are now written in the SwitchChunksLock() method since it's better placed there and the comment has been updated (it wasn't possible to put it there previously since it could thrown an exception): https://github.com/EventStore/EventStore/pull/3713/commits/1f7be5d0cfc215e2c3c0cb0a511e01f31d5071c8

Reply:
makes sense i think, im just making this unresolved for the minute to remind me to come back to it in a bit